### ISSUE TITLE
gpu: sdpa: split-barrier microkernel intermittently returns non-finite rows on Arc B580

### ISSUE THREAD

**Issue by haborite** [oneapi-src/oneDNN#5758]
# Summary

On Intel Arc B580, the oneDNN GPU fused SDPA implementation `ocl:micro:reusable` intermittently returns `NaN`/`Inf` output rows for finite f16 and bf16 inputs when used through PyTorch XPU on Windows with the SYCL/Level Zero runtime.

The same tensors remain finite with an unfused `matmul -> softmax -> matmul` control. Failed calls contain complete non-finite query-token rows; the non-finite element count is always a multiple of the head dimension (64). No error or warning is raised.

The failure reproduced with oneDNN v3.12.0 from the affected PyTorch wheels and with an unmodified oneDNN v3.12.3 statically linked into a local PyTorch build. However, standalone SYCL/Level Zero builds of the same oneDNN v3.12.0 and v3.12.3 revisions did not reproduce it in 20,000 f16 or 20,000 bf16 executions.

The same source revisions therefore produce different test results depending on the build or integration configuration. The oneDNN revision alone does not determine whether the failure occurs.

As an isolation experiment, all eight `intel_work_group_barrier_arrive` / `intel_work_group_barrier_wait` calls in `src/gpu/intel/sdpa/micro.cl` were replaced with full `barrier(CLK_LOCAL_MEM_FENCE)` calls. The resulting build still selected `ocl:micro:reusable` and produced zero non-finite calls in 100,000 f16 and 100,000 bf16 executions.

This strongly implicates the split-barrier SLM synchronization path and the conditions under which its GPU binary is built or executed. It does not by itself establish whether the lowest-level defect is in the oneDNN synchronization sequence, OpenCL compilation, SYCL/Level Zero runtime, the graphics driver, or another integration-dependent condition.

# Version

Released revisions compared in both standalone and PyTorch-integrated configurations:

```text
oneDNN v3.12.0
commit 80afa71049cd69a3df32adcccb623b12cd7baa22

oneDNN v3.12.3
commit 6db5f1ba860ca8f65c3f65c4498ad867e9642447
```

Verbose output from the v3.12.3 build:

```text
onednn_verbose,v1,info,oneDNN v3.12.3 (commit 6db5f1ba860ca8f65c3f65c4498ad867e9642447)
onednn_verbose,v1,info,gpu,runtime:SYCL
onednn_verbose,v1,info,gpu,engine,0,backend:Level Zero,name:Intel(R) Arc(TM) B580 Graphics,driver_version:1.15.38308,binary_kernels:enabled
onednn_verbose,v1,primitive,exec,gpu,sdpa,ocl:micro:reusable,forward_inference
```

Same-revision standalone comparison:

```text
oneDNN v3.12.0, commit 80afa71049cd69a3df32adcccb623b12cd7baa22
  PyTorch-integrated: reproduced
  standalone SYCL/Level Zero: 0/20000 f16, 0/20000 bf16

oneDNN v3.12.3, commit 6db5f1ba860ca8f65c3f65c4498ad867e9642447
  PyTorch-integrated: reproduced
  standalone SYCL/Level Zero: 0/20000 f16, 0/20000 bf16
```

The standalone builds used DPC++ 2026.1.1, the Level Zero backend, the same tensor dimensions and input distribution, and `ocl:micro:reusable`. They reused one primitive and one set of inputs and checked every output for non-finite elements after every execution.

Latest-`main` check:

```text
oneDNN main commit checked: 5274b630cd380291a88bf52dff598d53c83c2fd3
```

The split-barrier calls are still present in the forward SDPA microkernel at that revision. A standalone SYCL/Level Zero build of the earlier main revision `5bb5d33bfbb91e594d5fa1b82bd81b6d4569299d` also did not reproduce the non-finite output in 20,000 f16 or 20,000 bf16 executions.

Replacing v3.12.3 with `5274b630cd` inside the tested PyTorch revision was not a valid correctness comparison: the first SDPA call failed with `UR_RESULT_ERROR_DEVICE_LOST`, indicating an integration incompatibility. I therefore cannot claim that the intermittent non-finite output reproduces on the latest standalone `main` revision.

# Environment

```text
CPU: AMD Ryzen 7 2700X Eight-Core Processor
GPU: Intel(R) Arc(TM) B580 Graphics
OS: Microsoft Windows 11 Education 10.0.26200, 64-bit
GPU driver: 32.0.101.8864
Level Zero driver version reported by oneDNN: 1.15.38308
Python: CPython 3.13, 64-bit

Affected stable wheel:
  PyTorch: 2.13.0+xpu
  XPU runtime: 20260000
  oneDNN: v3.12.0, 80afa71049cd69a3df32adcccb623b12cd7baa22

Affected nightly wheel:
  PyTorch: 2.14.0.dev20260804+xpu
  XPU runtime: 20260100
  oneDNN: v3.12.0, 80afa71049cd69a3df32adcccb623b12cd7baa22

Local source build used for the v3.12.3 comparison:
  PyTorch: c0049254c669a7df79838e40c78e71584385a8aa
  oneDNN: v3.12.3, 6db5f1ba860ca8f65c3f65c4498ad867e9642447
  Build type: Release
  Host compiler: MSVC 19.44.35227.0
  SYCL compiler: Intel oneAPI DPC++/C++ 2026.1.0, build 20260617
  XPU AOT target: bmg
  CMake: 4.4.2
  Generator: Ninja
  DNNL_GPU_RUNTIME=SYCL
  DNNL_CPU_RUNTIME=NONE
  DNNL_LIBRARY_TYPE=STATIC
  ONEDNN_BUILD_GRAPH=ON

Standalone same-revision comparisons:
  oneDNN: v3.12.0 and v3.12.3
  SYCL compiler: Intel oneAPI DPC++/C++ 2026.1.1
  GPU runtime: SYCL
  Backend: Level Zero
  Device: Intel(R) Arc(TM) B580 Graphics
  Implementation: ocl:micro:reusable
  Test driver: locally instrumented benchdnn SDPA driver
```

# Steps to reproduce

The following script uses PyTorch only as the frontend that constructs and executes the oneDNN Graph SDPA primitive. No model or application framework is required.

Save as `repro.py`:

```python
import argparse

import torch
import torch.nn.functional as F
from torch.nn.attention import SDPBackend, sdpa_kernel


parser = argparse.ArgumentParser()
parser.add_argument("--dtype", choices=("float16", "bfloat16"), required=True)
parser.add_argument("--iterations", type=int, default=20000)
parser.add_argument("--math-iterations", type=int, default=1000)
args = parser.parse_args()

dtype = getattr(torch, args.dtype)
device = "xpu"
batch, heads, query_len, kv_len, head_dim = 8, 10, 4096, 77, 64

torch.manual_seed(0)
q = (torch.randn(batch, heads, query_len, head_dim) * 0.67).to(device, dtype)
k = torch.randn(batch, heads, kv_len, head_dim) * 3.2
k[:, :, 0, :] *= 30.0
k = k.to(device, dtype)
v = (torch.randn(batch, heads, kv_len, head_dim) * 0.95).to(device, dtype)

assert torch.isfinite(q).all().item()
assert torch.isfinite(k).all().item()
assert torch.isfinite(v).all().item()


def run(backend, iterations):
    failed_calls = 0
    bad_counts = []
    with sdpa_kernel(backend):
        for _ in range(iterations):
            out = F.scaled_dot_product_attention(
                q, k, v, dropout_p=0.0, is_causal=False
            )
            bad = (~torch.isfinite(out)).sum().item()
            if bad:
                failed_calls += 1
                bad_counts.append(bad)
    return failed_calls, bad_counts


fused_failed, fused_counts = run(SDPBackend.OVERRIDEABLE, args.iterations)
math_failed, math_counts = run(SDPBackend.MATH, args.math_iterations)

print(f"torch={torch.__version__}")
print(f"xpu={torch.version.xpu}")
print(f"device={torch.xpu.get_device_name(0)}")
print(f"dtype={dtype}")
print(f"fused_failed={fused_failed}/{args.iterations}")
print(f"fused_nonfinite_counts={fused_counts}")
print(f"math_failed={math_failed}/{args.math_iterations}")
print(f"math_nonfinite_counts={math_counts}")
```

Run both data types:

```powershell
python repro.py --dtype float16 --iterations 20000
python repro.py --dtype bfloat16 --iterations 20000
```

Verify that the fused oneDNN microkernel is selected:

```powershell
$env:ONEDNN_VERBOSE = "1"
python repro.py --dtype float16 --iterations 1 --math-iterations 0
```

The relevant verbose line should contain:

```text
gpu,sdpa,ocl:micro:reusable,forward_inference
```

Because the failure is intermittent, a single execution is not a sufficient correctness test. The observed rate varies between runs.

## Isolation patch

The following was applied to oneDNN v3.12.3 before rebuilding the static library and `torch_xpu.dll`:

```diff
diff --git a/src/gpu/intel/sdpa/micro.cl b/src/gpu/intel/sdpa/micro.cl
--- a/src/gpu/intel/sdpa/micro.cl
+++ b/src/gpu/intel/sdpa/micro.cl
@@
-        intel_work_group_barrier_arrive(CLK_LOCAL_MEM_FENCE);
+        barrier(CLK_LOCAL_MEM_FENCE);
@@
-        intel_work_group_barrier_wait(CLK_LOCAL_MEM_FENCE);
+        barrier(CLK_LOCAL_MEM_FENCE);
@@
-        intel_work_group_barrier_arrive(CLK_LOCAL_MEM_FENCE);
+        barrier(CLK_LOCAL_MEM_FENCE);
@@
-        intel_work_group_barrier_wait(CLK_LOCAL_MEM_FENCE);
+        barrier(CLK_LOCAL_MEM_FENCE);
@@
-        intel_work_group_barrier_arrive(CLK_LOCAL_MEM_FENCE);
+        barrier(CLK_LOCAL_MEM_FENCE);
@@
-        intel_work_group_barrier_wait(CLK_LOCAL_MEM_FENCE);
+        barrier(CLK_LOCAL_MEM_FENCE);
@@
-            intel_work_group_barrier_arrive(CLK_LOCAL_MEM_FENCE);
+            barrier(CLK_LOCAL_MEM_FENCE);
@@
-            intel_work_group_barrier_wait(CLK_LOCAL_MEM_FENCE);
+            barrier(CLK_LOCAL_MEM_FENCE);
```

These eight call sites form four logical synchronization pairs:

1. Q data written to and consumed from SLM;
2. publication and consumption of work-group-wide softmax maxima;
3. completion and consumption of S-tile stores; and
4. publication and consumption of the final column sums.

No tile, shape, dtype, or backend-selection change was made.

# Observed behavior

The central same-revision comparison is:

| oneDNN revision | Configuration | f16 non-finite calls | bf16 non-finite calls |
|---|---|---:|---:|
| v3.12.0 | PyTorch-integrated | 15/20,000 | 19/20,000 |
| v3.12.0 | Standalone SYCL/Level Zero | 0/20,000 | 0/20,000 |
| v3.12.3 | PyTorch-integrated | 2/20,000 in a representative run | 26/20,000 |
| v3.12.3 | Standalone SYCL/Level Zero | 0/20,000 | 0/20,000 |

The failure counts are stochastic and vary between runs. The table is not a performance comparison; it shows that the same oneDNN source revisions behaved differently under the two build/integration configurations.

Representative results from the affected v3.12.0 wheel:

```text
float16:
  fused_failed=15/20000
  math_failed=0/5000

bfloat16:
  fused_failed=19/20000
  math_failed=0/5000
```

The number of non-finite output elements in each failed call was always a multiple of 64. Representative bf16 counts included:

```text
128, 1024, 64, 128, 1024, 64, 256, 512
```

An unmodified local PyTorch build with oneDNN v3.12.3 also reproduced. Representative runs included 26 failing bf16 calls in 20,000 executions and 2 failing f16 calls in 20,000 executions. Separate QK matmul, softmax, and AV matmul controls remained finite.

Results after applying the full-barrier isolation patch while retaining `ocl:micro:reusable`:

```text
bfloat16: 0/100000 calls produced non-finite output
float16:  0/100000 calls produced non-finite output
```

Reinstalling the packaged patched build and retesting produced:

```text
bfloat16: 0/20000 calls produced non-finite output
float16:  0/20000 calls produced non-finite output
```

The full barriers may reduce overlap and performance. This is presented as an isolation result and conservative local workaround, not necessarily as the preferred upstream fix. It also does not prove that the source-level barrier sequence alone is defective, because changing it changes the generated GPU binary.

# Expected behavior

The fused SDPA primitive should always return finite output for these finite inputs. Repeated execution with identical inputs should not intermittently corrupt complete output rows.

# Related issues
- (closed) https://github.com/pytorch/pytorch/issues/192158

**Comment by umar456** [comment]
I was able to reproduce this issue. This seems like a bug in the driver/IGC and I need to isolate it to a specific version. @haborite can you try your test in a slightly older driver? It seems to be working with a newer version but it is unreleased.

**Comment by haborite** [comment]
This issue was reproduced with all the drivers I tested.

## Fixed test configuration

- GPU: Intel Arc B580
- OS: Windows 11
- PyTorch: `2.14.0.dev20260804+xpu`
- Bundled oneDNN: v3.12.0 (`80afa71049cd69a3df32adcccb623b12cd7baa22`)
- Backend: SYCL / Level Zero
- oneDNN implementation: `gpu,sdpa,ocl:micro:reusable`
- Tensors: Q `[8,10,4096,64]`, K/V `[8,10,77,64]`
- Inputs: fixed seed and identical generation procedure for every driver
- Test: unpatched `SDPBackend.OVERRIDEABLE`, 20,000 executions per dtype
- Cache: a separate `SYCL_CACHE_DIR` for every driver version
- Driver installation: Intel installer with `-o -f -s`, followed by a reboot

No split-barrier workaround or SDPA fallback was enabled.

## Results

| Intel graphics driver | Level Zero runtime driver | bf16 failed calls | fp16 failed calls | Result |
|---|---:|---:|---:|---|
| 32.0.101.8864 | 1.15.38308 | 4 / 20,000 | 16 / 20,000 | reproduced |
| 32.0.101.8861 | 1.15.38308 | 6 / 20,000 | 4 / 20,000 | reproduced |
| 32.0.101.8824 | 1.15.37858 | 3 / 20,000 | 3 / 20,000 | reproduced |
| 32.0.101.8735 | 1.15.37669 | 6 / 20,000 | 5 / 20,000 | reproduced |

Representative non-finite element counts:

- 8864 bf16: `1024, 64, 64, 64`
- 8864 fp16: `192, 512, 1152, 448, 512, 384, 384, 512, 192, 704, 512, 448, 256, 1024, 512, 256`
- 8861 bf16: `1024, 64, 1024, 256, 1024, 1024`
- 8861 fp16: `256, 1600, 448, 1024`
- 8824 bf16: `64, 1024, 1024`
- 8824 fp16: `832, 384, 640`
- 8735 bf16: `64, 2048, 1024, 1024, 261, 64`
- 8735 fp16: `1024, 384, 768, 384, 1600`

The `261`-element bf16 failure on 8735 is an exception to the complete-row/multiple-of-64 pattern seen in the other recorded failures.

**Comment by haborite** [comment]
## Follow-up: 32.0.101.8974

Retested after installing the latest driver: 32.0.101.8974 (WHQL).

- Windows driver: 32.0.101.8974 (2026-08-11)
- Level Zero runtime driver: `1.15.39183+1` (`1.15.39183` in oneDNN verbose)
- Selected implementation: `gpu,sdpa,ocl:micro:reusable`
- oneDNN: v3.12.0 (`80afa71049cd69a3df32adcccb623b12cd7baa22`)
- bf16: 2 / 20,000 failed calls; non-finite counts `128, 128`
- fp16: 4 / 20,000 failed calls; non-finite counts `320, 320, 512, 320`

The failure remains reproducible on 32.0.101.8974. All recorded non-finite counts in this run were multiples of the 64-element head dimension.
